### PR TITLE
Google Summer of Code 2016 JavaScript Testing project

### PR DESCRIPTION
**Pull request by Ruchiranga**
#### Summary of Changes

The set of custom JavaScript libraries in Joomla does not have any tests. This PR contains a test suite that will be run, like the PHP unit tests, automatically for every pull request and commit with Travis. 

Shown below are some screenshots of the Travis log with the JavaScript test results.

core.js test results

validate.js test results

repeatable.js test results

The Travis log will also show an overview of the test coverage achieved.

Test coverage and overview

As testing framework we use Jasmine, which is run by Karma-Test Runner in a real Firefox browser.

Core JavaScript Libraries covered by this Pull Request:
1. caption.js
2. core.js
3. permissions.js
4. repeatable.js
5. validate.js
#### Documentation

The documentation is currently available [here](https://docs.google.com/document/d/1ZTzJiXwh4lNI9pPT38WqBy28nCd1duAD5fQakxZRm7s/edit). This will be moved into the Joomla! documentation soon.
#### Testing Instructions

None, since this is only run on Travis. If you want to run it locally, please take a look at the documentation.
#### Demonstration

Demonstration video can be watched via this [link](https://www.youtube.com/watch?v=Tp_mLqMhRuA).

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,18 +1,19 @@
-# Forces new Travis-CI Infrastructure
-sudo: false
```
Comment (mbabker) [inline .travis.yml]: Is this needed as it's going to affect all jobs?

Comment (yvesh) [inline .travis.yml]: Not sure, why it's in there .. I think from the beginning of GSoC project, where it still was that way in core.. Going to reset it back. 

@mbabker For the unit tests sudo seems not to be needed, any reasons it was reenabled?

**Comment by Ruchiranga** [comment]
The Travis log for the JavaScript tests in this PR is available [here ](https://travis-ci.org/joomla/joomla-cms/jobs/149521735).
The JavaScript tests are run only on the build job shown below.
JS Tests Build Job

File: .travis.yml (hunk 1 above)
Comment (mbabker) [inline .travis.yml]: We turned it off to use the newer (faster) Travis-CI structure by default.  So my only guess is someone changed this testing something along the way.

**Comment by zero-24** [comment]
Good work @Ruchiranga Looks really nice. Thanks to the GSOC Team!

**Comment by Ruchiranga** [comment]
Thank you @zero-24 :)

**Comment by gunjanpatel** [comment]
Good work @Ruchiranga

**Comment by roland-d** [comment]
@Ruchiranga That looks very cool. The video and your screenshots show when everything goes right. Can you post some shots of what happens when I break something with a code change?

**Comment by Ruchiranga** [comment]
@roland-d 
Here is a screenshot of the `Joomla.submitform` function in core.js.
Joomla.submitform before

And all the tests covering that function are passing.
Joomla.submitform before log

Now here, I comment line `form.setAttribute('novalidate', !validate)` mimicking a situation where we have mistakenly deleted that line of code.
Joomla.submitform after code

As a result, the test case covering that line would fail. It will be shown in the log as follows.
Joomla.submitform after log

The overall summary of the tests would also show that one test case is failing.
Joomla.submitform after log summary

**Comment by Ruchiranga** [comment]
Thank you @gunjanpatel :)

**Comment by roland-d** [comment]
@Ruchiranga Thank you for that explanation looks good. Good job.

**Comment by yvesh** [comment]
I have tested this item :white_check_mark: successfully on fda4cdbfb78c2ed9a78aaeca08a5426f0f49c6c7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11420">issues.joomla.org/joomla-cms/11420</a>.</sub>

**Comment by rdeutz** [comment]
Tested with travis and in an local Environment. I will merge on Review <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11420">issues.joomla.org/joomla-cms/11420</a>.</sub>

**Comment by rdeutz** [comment]
Good job @Ruchiranga! One thing we might can add is that it is not needed that phpunit runs for the javascript tests so if we can exclude this we can have the result faster. No big deal at all.

**Comment by Ruchiranga** [comment]
@rdeutz Thank you! Yes I agree. We will make that change. Thank you very much again for taking your time to go through this and get it merged :)

**Pull request merged**